server/interceptor/interceptor.go

```diff
@@ -107,24 +109,24 @@ func getRPCInfo(req interface{}, fullMethod, quotaUser string) (*rpcInfo, error)
 		return nil, status.Errorf(codes.Internal, "cannot retrieve treeID from request: %T", req)
 	}
 
-	serviceName, methodName, err := parseFullMethod(fullMethod)
-	if err != nil {
-		return nil, err
-	}
-
 	var treeType trillian.TreeType
-	switch serviceName {
-	case "trillian.TrillianAdmin":
+	switch {
+	case isAdminRPC(req):
 		treeType = trillian.TreeType_UNKNOWN_TREE_TYPE // unrestricted
-	case "trillian.TrillianLog":
+	case isLogRPC(req):
 		treeType = trillian.TreeType_LOG
-	case "trillian.TrillianMap":
+	case isMapRPC(req):
 		treeType = trillian.TreeType_MAP
 	default:
-		return nil, status.Errorf(codes.Internal, "cannot determine treeType for request: %T", req)
+		return nil, status.Errorf(codes.Internal, "unmapped request type: %T", req)
 	}
 
-	readonly := strings.HasPrefix(methodName, "Get") || strings.HasPrefix(methodName, "List")
+	reqType := fmt.Sprintf("%T", req)
```

Comment:
I don't see that this approach has significantly less human intervention than what I suggest -- in either case you need to add an entry to an existing list, whether it's:
```go
func isAdminRPC(req interface{}) bool {
	switch req.(type) {
	// Please keep in alphabetical order
	case *trillian.CreateTreeRequest:
	case *trillian.DeleteTreeRequest:
	case *trillian.GetTreeRequest:
	case *trillian.ListTreesRequest:
	case *trillian.UpdateTreeRequest:
	default:
		return false
	}
	return true
}
```
or:
```go
func isAdminRPC(req interface{}) (bool, bool) {
	isAdmin := false
	readonly := false
	switch req.(type) {
	case *trillian.CreateTreeRequest, *trillian.DeleteTreeRequest, *trillian.UpdateTreeRequest:
		isAdmin = true
		readonly = false
	case *trillian.GetTreeRequest, *trillian.ListTreesRequest:
		isAdmin = true
		readonly = true
	}
	return isAdmin, readonly
}
```
is hardly a big deal (and note I'm not suggesting a `treeID` mapping, so that's a red herring).  

What is more significant for maintainability is that the convention that `List..` / `Get...` methods are readonly is more explicit (and greppable), and less likely to go wrong if someone foolishly chooses the wrong name.

Fundamentally, I think that doing reflection and string matching on the names of types should be a last resort, reserved for when there's no other way to get the same effect.